Switch to using BLDDIR as discussed in #3

Thyrum opened the pull request:
Closes #3.

avdstaaij committed:
Change make clean to not remove dirs inside BINDIR

This change was discussed in #3.

Previously, make clean would recursively remove empty directories inside
BINDIR. This was too intrusive, as it could cause problems if the user
had set BINDIR to ".".
Now, only BINDIR itself is deleted (if it exists and is empty). This
change is not possible for OBJDIR and DEPDIR, since the build process
may add directory trees in them to reflect the source file directory
tree.

avdstaaij commented on Makefile:
```diff
@@ -54,8 +54,7 @@ BINNAME_D = $(BINNAME_R)_debug
-DEPDIR = build/dep
+BLDDIR = build
 
 SUBDIR_R = release
 SUBDIR_D = debug
```
If we hardcode the names ```obj``` and ```dep```, shouldn't we also hardcode these? (That is, move these lines below the "Are you sure ..."-header.)

Thyrum replied:
Agreed. Since the the variables are irrelevant for the (normal) user experience and a normal user will never have to change them, they can and should be hardcoded.

avdstaaij replied:
I have hardcoded them.

avdstaaij commented on Makefile:
```diff
@@ -73,13 +72,22 @@ MKBIN_D = $(CXX) $(CXXFLAGS) $(FLAGS_D) $^ $(LDFLAGS) -o $@
 
+# Checks
+
+ifeq ($(BLDDIR), .)
+  $(error "You MONSTER, never set BLDDIR to '.'. This clutters the directories and complicates cleaning.")
+endif
```
The "and complicates cleaning" part is not really a good argument anymore, since our Makefile would handle it perfectly fine now, thanks to the hardcoded subdirectories ```obj``` and ```dir```. I would remove that part.

In fact, I just realized that the only reason to keep this test is to prevent the Makefile from removing empty directories in user-created ```obj``` and ```dir``` directories at the project root, an _extremely_ niche safety measure. We could remove the test entirely for simplicity, though it is funny.

Thyrum replied:
Even though I like the idea of insulting the user, I agree setting `BLDDIR` to `.` doesn't create a big enough problem to warrant an error message. I'm fine with removing this check.

avdstaaij replied:
I changed the warning, but left in the check because it's just too funny.

avdstaaij reviewed (commented):
I have made some changes and left some comments. Please take a look.

avdstaaij committed:
Allow make clean to clean any BINDIR

Before, BINDIR was only cleaned correctly if BINDIR was a relative path
from the Makefile in the simplest possible form (e.g. no './' prefix).
Now, BINDIR can be any directory. It can be both relative and absolute,
and it no longer needs to be in the simplest form.

avdstaaij committed:
Make BLDDIR='.' warning more accurate

This check is really just for fun, but the warning is now more accurate.
The warning doesn't need a trailing period because make already adds one.

avdstaaij reviewed (commented):
I think this can be merged. @Thyrum, do you agree?

Thyrum committed:
Switch to using BLDDIR as discussed in #3

avdstaaij committed:
Update Makefile instructions

avdstaaij committed:
Set default BLDDIR to "build"

avdstaaij committed:
Move SUBDIR_R/D out of user settings

Since OBJDIR and DEPDIR were moved out of the settings, it made sense to
move these out as well.
Normal users will never have to change these variables.

avdstaaij committed:
Fix make clean warning when BINDIR is "."

rmdir always prints a warning if "." is passed as argument, even with
--ignore-fail-on-non-empty. This would cause a warning to be printed
when using make clean while BINDIR is ".". This change fixes that.
